tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <sys/socket.h>
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_socket.h"
+
+/*
+ * sendmsg and recvmsg are syscalls which can be used to send 'real' data along
+ * with ancillary data. Ancillary data is used to communicate to the socket the
+ * type of the TLS record being sent/received.
+ *
+ * Ancillary data macros (CMSG_*) are platform specific and gated.
+ */
+
+S2N_RESULT s2n_ktls_send_msg_impl(int sock, struct msghdr *msg,
+        const struct iovec *msg_iov, s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wcast-qual"
+    /* set send buffer */
+    msg->msg_iov = (struct iovec *) msg_iov;
+#pragma GCC diagnostic pop
```

Comment:
Its going to eventually be called from https://github.com/aws/s2n-tls/blob/main/tls/s2n_send.c#L109 which defines the argument as `const struct iovec *bufs`

Reply:
Casting a const to a non-const is still extremely dangerous. I would do it as early as possible, probably meaning in s2n_send. The deeper in the stack you do it, the more likely it is to cause unexpected errors because someone assumed it was const.

Reply:
hmm, thought later in the stack would be safer. I can make it earlier though.

`msg->msg_iov` expects a non-const so we do need to do the conversion.

Reply:
removed const and will do conversion in s2n_send https://github.com/aws/s2n-tls/pull/4073/commits/3dd57a302063e871b1f44639747c528bd03c52ee